codelab-master/apps/playground/src/app/quiz/shared/services/quiz.service.ts

```diff
@@ -0,0 +1,217 @@
+import { Injectable } from '@angular/core';
+import { ActivatedRoute, Router } from '@angular/router';
+import { BehaviorSubject } from 'rxjs';
+import { Howl } from 'howler';
+
+import { QUIZ_DATA } from '@codelab-quiz/shared/quiz-data';
+import { Option, Quiz, QuizQuestion } from '@codelab-quiz/shared/models/';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class QuizService {
+  quizData: Quiz[] = JSON.parse(JSON.stringify(QUIZ_DATA));
+  question: QuizQuestion;
+  questions: QuizQuestion[];
+  answers: number[];
+  multipleAnswer: boolean;
+  totalQuestions: number;
+  currentQuestionIndex = 1;
+
+  quizId: string;
+  startedQuizId: string;
+  continueQuizId: string;
+  completedQuizId: string;
+  indexOfQuizId: number;
+  quizCompleted: boolean;
+  status: string;
+
+  correctAnswers = [];
+  correctAnswersForEachQuestion = [];
+  correctAnswerOptions: number[] = [];
+  numberOfCorrectAnswers: number;
+  correctAnswersCountSubject = new BehaviorSubject<number>(0);
+
+  userAnswers = [];
+  previousUserAnswers: any[] = [];
+  previousUserAnswersText: any[] = [];
+  previousUserAnswersInnerText = [];
+  previousUserAnswersTextSingleAnswer: string[] = [];
+  previousUserAnswersTextMultipleAnswer: string[] = [];
+
+  explanation: string;
+  explanationText: string;
+  correctOptions: string;
+  correctMessage: string;
+
+  isAnswered: boolean;
+  checkedShuffle: boolean;
+
+  correctSound = new Howl({
+    src: '../../../assets/audio/sound-correct.mp3',
+    html5: true,
+    format: ['mp3', 'aac']
+  });
+  incorrectSound = new Howl({
+    src: '../../../assets/audio/sound-incorrect.mp3',
+    html5: true,
+    format: ['mp3', 'aac']
+  });
+
+
+  constructor(
+    private activatedRoute: ActivatedRoute,
+    private router: Router
+  ) {
+    this.quizId = this.activatedRoute.snapshot.paramMap.get('quizId');
+    this.indexOfQuizId = this.quizData.findIndex(el => el.quizId === this.quizId);
+  }
+
+  getCorrectAnswers(question: QuizQuestion) {
+    if (this.question) {
+      const identifiedCorrectAnswers = question.options.filter((option) => option.correct);
+      this.numberOfCorrectAnswers = identifiedCorrectAnswers.length;
+      this.correctAnswerOptions = identifiedCorrectAnswers.map((option) => question.options.indexOf(option) + 1);
+
+      this.correctAnswersForEachQuestion.push(this.correctAnswerOptions);
+      this.correctAnswers.push(this.correctAnswersForEachQuestion);
+
+      this.setCorrectAnswerMessagesAndExplanationText(this.correctAnswersForEachQuestion.sort());
+      return identifiedCorrectAnswers;
+    }
+  }
+
+  // shuffle questions array in-place using Durstenfeld's shuffling algorithm
+  shuffleQuestions(questions: QuizQuestion[]): void {
+    for (let i = questions.length - 1; i >= 0; i--) {
+      const j = Math.floor(Math.random() * (i + 1));
+      [questions[i], questions[j]] = [questions[j], questions[i]];
+    }
+  }
+
+  // shuffle answers array in-place using Durstenfeld's shuffling algorithm
```

Comment:
awesome!

Reply:
Yes I'll try to combine the two functions together.

Reply:
Yes I used a generic function and called it, seems to be working fine.